### ISSUE TITLE
seL4_ARM_Page_Invalidate_Data doesn't check if the page is writeable (ARMv8)

### ISSUE THREAD

**Issue by oscarwcl** [seL4/seL4#214]
Calling `seL4_ARM_Page_Invalidate_Data` on a page cap that is mapped without write permissions will cause a kernel fault:
```
KERNEL DATA ABORT!
Faulting instruction: 0xffffff8001011d08
FAR: 0x400000 ESR (DFSR): 0x9600014f
halting...
Kernel entry via Syscall, number: 1, Call
Cap type: 1, Invocation tag: 49
```
The faulting instruction is the `dc ivac` instruction here:
https://github.com/seL4/seL4/blob/1adb7e9b9162902b3f9868232827b90456d04d53/include/arch/arm/arch/64/mode/machine.h#L291-L295
which requires write permission to the vaddr. I traced the function calls from here back to `decodeARMFrameInvocation` and as far as I can tell there isn't anything that checks the permissions on the page.

**Comment by yyshen** [comment]
@oscarwcl good catch!

**Comment by yyshen** [comment]
@kent-mcleod @lsf37 What I have in mind to fix this is to use "address translation" instruction to check the write permission of the address before performing an invalidation, unless the the access right is stored somewhere (other than in the page table entries) for a quick check.

**Comment by lsf37** [comment]
The ultimate right (including masks etc) is only in the page table entries, so I guess we do have to walk the table, yes. Is the plan to put the instruction directly inside `invalidateByVA`? (That would probably be convenient for the proofs).

I need to double-check the `Page_Invalidate_Data` invocation if it can only be called on the current page table, or on any frame cap in any table. If the latter, we need to walk the right table..

Is this requirement for write permission new in aarch64, or have we missed this in Armv7 and before as well?

**Comment by lsf37** [comment]
Ah, for flushing we're setting the address space to be flushed to be current for the flush anyway, so that part is all good.

**Comment by yyshen** [comment]
> The ultimate right (including masks etc) is only in the page table entries, so I guess we do have to walk the table, yes. Is the plan to put the instruction directly inside `invalidateByVA`? (That would probably be convenient for the proofs).

It depends if an error would be returned to the caller. If we could simply ignore invalidate without write permission, putting the check in invalidateByVA is OK. But this is somewhat inefficient, since the function is executed for each cache line, but the permission is for a page.

> 
> I need to double-check the `Page_Invalidate_Data` invocation if it can only be called on the current page table, or on any frame cap in any table. If the latter, we need to walk the right table..

The use of "AT" instruction is to let the hardware to perform the page table walk for the caller, saving the code for walking the table.

> 
> Is this requirement for write permission new in aarch64, or have we missed this in Armv7 and before as well?

Well, it is more complicated since for some armv7 platforms (imx7), we also have the code for L2 cache controller. I forgot the details, but will check.

**Comment by lsf37** [comment]
> It depends if an error would be returned to the caller. If we could simply ignore invalidate without write permission, putting the check in invalidateByVA is OK. But this is somewhat inefficient, since the function is executed for each cache line, but the permission is for a page.

Right, once for each cache line would not be good. It makes sense to add a new machine interface operation then.

The problem with returning an error to the user is that this would have to be in `decode`, and we'd then have a machine instruction that determines the result of of the decode, which means we need the precise semantics of that instruction. Could be done, but is work and has the risk that we didn't capture the instruction correctly. How bad would it be to walk the table in software?

Semantically, returning an error sounds like the cleanest solution, though. @kent-mcleod do you have an opinion on this?

**Comment by yyshen** [comment]
> > It depends if an error would be returned to the caller. If we could simply ignore invalidate without write permission, putting the check in invalidateByVA is OK. But this is somewhat inefficient, since the function is executed for each cache line, but the permission is for a page.
> 
> Right, once for each cache line would not be good. It makes sense to add a new machine interface operation then.
> 
> The problem with returning an error to the user is that this would have to be in `decode`, and we'd then have a machine instruction that determines the result of of the decode, which means we need the precise semantics of that instruction. Could be done, but is work and has the risk that we didn't capture the instruction correctly. How bad would it be to walk the table in software?

In terms of implementation, it should be fine since we already have lookupFrame and related functions. But of course the kernel does not cache the intermediate lookup results, this translates to a full walk for each page included in the [start, start + size). If the size is big, the lookups are expensive, unless we also put a limit on the size. For instance, we only allow the system call to invalid 10 pages each time at max. The AT instruction just uses the results cached in TLB, unless there is a miss, I think.
 
> 
> Semantically, returning an error sounds like the cleanest solution, though. @kent-mcleod do you have an opinion on this?

**Comment by kent-mcleod** [comment]
It may be time to consider using an alternative approach for implementing the cache maintenance operations.  The `seL4_ARM_Page_Invalidate_Data` invocation potentially still needs to work even if the mapping the Frame cap refers to doesn't have the write attribute.  Consider a scenario where a cached read only mapping is used to receive data from a device that writes directly to memory.  The invalidation operation would still be required each time the driver updates the memory.
Another issue is that if there is a shared page between two software components, one with read access, the other with read/write access.  Should the component with only read access be able to perform an invalidate on the frame which would invalidate any writes by the other component?  Should the kernel actually restrict cache maintenance operations via a different access right?

**Comment by yyshen** [comment]
> It may be time to consider using an alternative approach for implementing the cache maintenance operations. The `seL4_ARM_Page_Invalidate_Data` invocation potentially still needs to work even if the mapping the Frame cap refers to doesn't have the write attribute. Consider a scenario where a cached read only mapping is used to receive data from a device that writes directly to memory. The invalidation operation would still be required each time the driver updates the memory.

> Another issue is that if there is a shared page between two software components, one with read access, the other with read/write access. Should the component with only read access be able to perform an invalidate on the frame which would invalidate any writes by the other component?

My guess this is why the invalidate needs write permission, but clean and invalidate does not. You do not disturb others' writes unless you can also write.

> Should the kernel actually restrict cache maintenance operations via a different access right?

**Comment by gernotheiser** [comment]
Not sure I have been following all this correctly, but any sort of cache management is a privilege. If anyone could do it, they can impact performance of other threads, and cache flushes can also be used to implement timing channels. Obviously invalidating (without write-back) needs write permission.

**Comment by lsf37** [comment]
The operation currently requires a FrameCap that covers the area to be flushed. So that part was fine. It did not require further permissions (`write`), though, which is a problem.

I think there are two issues: a) how to implement a check for permission (is checking the cap enough or do we need to walk the page table) and b) which permission is required.

@kent-mcleod, you have a point that always requiring `write` for a flush may be too much -- although in the scenario that you described (mapped read-only, but a device can write to it), wouldn't we get the same fault we are getting now? If the hardware requires the page to be mapped with `write` for this operation, there's not much choice for us apart from requiring it too. Or am I missing something? 

As for the implementation of a check for write permission: checking the cap is not good enough if the requirement is that the page table entry is mapped `write`, because a cap with `write` permission is allowed to map `read-only`. So we do have to walk the table to find out. I'd think we have to walk it only once, though. Even for a larger page size, the mapping rights for all PTEs involved have to be the same (if there are multiple).

**Comment by yyshen** [comment]
> The operation currently requires a FrameCap that covers the area to be flushed. So that part was fine. It did not require further permissions (`write`), though, which is a problem.

This is seL4_ARM_VSpace_Invalidate_Data(vspace, start, end), which takes a top level page table cap as the first argument.

> 
> I think there are two issues: a) how to implement a check for permission (is checking the cap enough or do we need to walk the page table) and b) which permission is required.
> 
> @kent-mcleod, you have a point that always requiring `write` for a flush may be too much -- although in the scenario that you described (mapped read-only, but a device can write to it), wouldn't we get the same fault we are getting now? If the hardware requires the page to be mapped with `write` for this operation, there's not much choice for us apart from requiring it too. Or am I missing something?
> 
> As for the implementation of a check for write permission: checking the cap is not good enough if the requirement is that the page table entry is mapped `write`, because a cap with `write` permission is allowed to map `read-only`. So we do have to walk the table to find out. I'd think we have to walk it only once, though. Even for a larger page size, the mapping rights for all PTEs involved have to be the same (if there are multiple).

**Comment by lsf37** [comment]
> This is seL4_ARM_VSpace_Invalidate_Data(vspace, start, end), which takes a top level page table cap as the first argument.

Right, I've been staring at the wrong operation then :-). I thought we were looking at `seL4_ARM_Page_Invalidate_Data`, but yes `seL4_ARM_VSpace_Invalidate_Data` would have the same problem, only worse. Would we have to check that the entire VSpace has only writable mappings if the user requests `[0,USER_TOP]`? That sounds daunting and like a potential denial of service.

**Comment by yyshen** [comment]
> > This is seL4_ARM_VSpace_Invalidate_Data(vspace, start, end), which takes a top level page table cap as the first argument.
> 
> Right, I've been staring at the wrong operation then :-). I thought we were looking at `seL4_ARM_Page_Invalidate_Data`, but yes `seL4_ARM_VSpace_Invalidate_Data` would have the same problem, only worse. Would we have to check that the entire VSpace has only writable mappings if the user requests `[0,USER_TOP]`? That sounds daunting and like a potential denial of service.

We only need to check the requested range [start, end). This is where we could put a size limit on the operation: for each invocation, the kernel only processes N pages at max, and return an error code to indicate that.

**Comment by yyshen** [comment]
> Not sure I have been following all this correctly, but any sort of cache management is a privilege. If anyone could do it, they can impact performance of other threads, and cache flushes can also be used to implement timing channels. Obviously invalidating (without write-back) needs write permission.

Yes. The system calls are provided instead of enabling EL0 access to cache maintenance instructions. Caps are required to perform the operations.

**Comment by lsf37** [comment]
> We only need to check the requested range [start, end). This is where we could put a size limit on the operation: for each invocation, the kernel only processes N pages at max, and return an error code to indicate that.

I guess we could do that. Question is what those operations are usually used for and how annoying such a restriction would be on user level. I.e. if this mostly used for something the size of a page, then it's no problem, but if someone wants to regularly use it for large portions of the entire address space, it would probably be too much.

It sounds a bit ad hoc, but it would solve the problem. We could also make the operation preemptable, and split it up inside the kernel. More work, but also doable.

> Alternatively, we do not check the permission at all, but we handle the kernel exception in the exception handler by killing the calling thread and resuming execution. However, I guess it is harder for verification to handle this.

Yes, that would break the current execution model of not taking any exceptions in the kernel. We could change that if we wanted to, it wouldn't necessarily be harder, but it would make a stronger assumptions about execution than we do now, i.e. our binary verification wouldn't catch any mistakes we're making in that, because it also assumes no exceptions. Then again, we sort-of do that already for the interrupt we might be taking at the preemption point. It's not really much worse than that. 

The main difference with taking the exception would be that the error does not happen in decode, but during the operation, and you'd potentially have a half-finished operation that then fails in the middle on access rights violation. We're usually trying to avoid that.

**Comment by yyshen** [comment]
> > We only need to check the requested range [start, end). This is where we could put a size limit on the operation: for each invocation, the kernel only processes N pages at max, and return an error code to indicate that.
> 
> I guess we could do that. Question is what those operations are usually used for and how annoying such a restriction would be on user level. I.e. if this mostly used for something the size of a page, then it's no problem, but if someone wants to regularly use it for large portions of the entire address space, it would probably be too much.

I could not imagine a regular application would invalidate the entire address regularly ... My feeling for the VSpace invalidate syscall is more about convenience, so the users do not to find the FrameCaps behind the virtual addresses for invalidation, if the users have the root vspace caps. It is a bit ad-hoc, but a reasonable restriction.

> 
> It sounds a bit ad hoc, but it would solve the problem. We could also make the operation preemptable, and split it up inside the kernel. More work, but also doable.

Yep doable. Another question is if the application failed the permission check should be terminated or be notified with an error code. We could also call the fault handler of the application.

> 
> > Alternatively, we do not check the permission at all, but we handle the kernel exception in the exception handler by killing the calling thread and resuming execution. However, I guess it is harder for verification to handle this.
> 
> Yes, that would break the current execution model of not taking any exceptions in the kernel. We could change that if we wanted to, it wouldn't necessarily be harder, but it would make a stronger assumptions about execution than we do now, i.e. our binary verification wouldn't catch any mistakes we're making in that, because it also assumes no exceptions. Then again, we sort-of do that already for the interrupt we might be taking at the preemption point. It's not really much worse than that.
> 
> The main difference with taking the exception would be that the error does not happen in decode, but during the operation, and you'd potentially have a half-finished operation that then fails in the middle on access rights violation. We're usually trying to avoid that.

Well, this sounds more fun to me :-)

**Comment by yyshen** [comment]
@lsf37 Which one of the three choices you would like to try out? PR #215 disables EL0 cache maintenance instructions explicitly.

**Comment by lsf37** [comment]
Not sure yet. @kent-mcleod do you have any opinions?

**Comment by kent-mcleod** [comment]
> Not sure I have been following all this correctly, but any sort of cache management is a privilege. If anyone could do it, they can impact performance of other threads, and cache flushes can also be used to implement timing channels. Obviously invalidating (without write-back) needs write permission.

@gernotheiser I agree that invalidating needs some sort of write permission to the cache because it's modifying the cache's state, but I was trying to point out that inheriting the write permission from the underlying frame doesn't always make sense.  In cases where the component should only have read access to the underlying memory it may still require write access to its cache to perform invalidation but you can't achieve that at the moment.  Any scenario where another component is writing to the underlying memory but is outside of the cache hierarchy of the reading component would have this issue.

> @kent-mcleod, you have a point that always requiring write for a flush may be too much -- although in the scenario that you described (mapped read-only, but a device can write to it), wouldn't we get the same fault we are getting now? If the hardware requires the page to be mapped with write for this operation, there's not much choice for us apart from requiring it too. Or am I missing something?

@lsf37 The device is either accessing memory via bypassing the mmu, or through an IOMMU that may have different mapping permissions for the memory.

> Right, I've been staring at the wrong operation then :-). I thought we were looking at seL4_ARM_Page_Invalidate_Data, but yes seL4_ARM_VSpace_Invalidate_Data would have the same problem, only worse. Would we have to check that the entire VSpace has only writable mappings if the user requests [0,USER_TOP]? That sounds daunting and like a potential denial of service.

The VSpace operations are still only allowed to operate one page at a time, and their implementation looks up the frame cap to validate that something is mapped there.

My opinion is that we find a way to place the actual mapping information into the frame cap and as part of decode reject the operation as invalid if the mapping doesn't have write rights.  There is space on aarch64 but not on aarch32, we could reduce the ASID space by 1 bit and this should be enough to encode whether the page is mapped with write rights.
- we shouldn't cause an exception for an unexceptional condition that we could avoid if we track the state ourselves.  
- Longer term we should find a different way of implementing cache maintenance operations so they more accurately reflect the underlying hardware architecture.

**Comment by yyshen** [comment]
> > Not sure I have been following all this correctly, but any sort of cache management is a privilege. If anyone could do it, they can impact performance of other threads, and cache flushes can also be used to implement timing channels. Obviously invalidating (without write-back) needs write permission.
> 
> @gernotheiser I agree that invalidating needs some sort of write permission to the cache because it's modifying the cache's state, but I was trying to point out that inheriting the write permission from the underlying frame doesn't always make sense. In cases where the component should only have read access to the underlying memory it may still require write access to its cache to perform invalidation but you can't achieve that at the moment. Any scenario where another component is writing to the underlying memory but is outside of the cache hierarchy of the reading component would have this issue.
> 
> > @kent-mcleod, you have a point that always requiring write for a flush may be too much -- although in the scenario that you described (mapped read-only, but a device can write to it), wouldn't we get the same fault we are getting now? If the hardware requires the page to be mapped with write for this operation, there's not much choice for us apart from requiring it too. Or am I missing something?
> 
> @lsf37 The device is either accessing memory via bypassing the mmu, or through an IOMMU that may have different mapping permissions for the memory.
> 
> > Right, I've been staring at the wrong operation then :-). I thought we were looking at seL4_ARM_Page_Invalidate_Data, but yes seL4_ARM_VSpace_Invalidate_Data would have the same problem, only worse. Would we have to check that the entire VSpace has only writable mappings if the user requests [0,USER_TOP]? That sounds daunting and like a potential denial of service.
> 
> The VSpace operations are still only allowed to operate one page at a time, and their implementation looks up the frame cap to validate that something is mapped there.

@lsf37 My apology here. I  just looked at the code; there is check in the kernel to ensure the start and end addresses are in the same page in the VSpace flush function. So, we do not have a denial of service issue. 

> 
> My opinion is that we find a way to place the actual mapping information into the frame cap and as part of decode reject the operation as invalid if the mapping doesn't have write rights. There is space on aarch64 but not on aarch32, we could reduce the ASID space by 1 bit and this should be enough to encode whether the page is mapped with write rights.
> 
> * we shouldn't cause an exception for an unexceptional condition that we could avoid if we track the state ourselves.
> * Longer term we should find a different way of implementing cache maintenance operations so they more accurately reflect the underlying hardware architecture.

**Comment by yyshen** [comment]
> > > Not sure I have been following all this correctly, but any sort of cache management is a privilege. If anyone could do it, they can impact performance of other threads, and cache flushes can also be used to implement timing channels. Obviously invalidating (without write-back) needs write permission.
> > 
> > 
> > @gernotheiser I agree that invalidating needs some sort of write permission to the cache because it's modifying the cache's state, but I was trying to point out that inheriting the write permission from the underlying frame doesn't always make sense. In cases where the component should only have read access to the underlying memory it may still require write access to its cache to perform invalidation but you can't achieve that at the moment. Any scenario where another component is writing to the underlying memory but is outside of the cache hierarchy of the reading component would have this issue.
> > > @kent-mcleod, you have a point that always requiring write for a flush may be too much -- although in the scenario that you described (mapped read-only, but a device can write to it), wouldn't we get the same fault we are getting now? If the hardware requires the page to be mapped with write for this operation, there's not much choice for us apart from requiring it too. Or am I missing something?
> > 
> > 
> > @lsf37 The device is either accessing memory via bypassing the mmu, or through an IOMMU that may have different mapping permissions for the memory.
> > > Right, I've been staring at the wrong operation then :-). I thought we were looking at seL4_ARM_Page_Invalidate_Data, but yes seL4_ARM_VSpace_Invalidate_Data would have the same problem, only worse. Would we have to check that the entire VSpace has only writable mappings if the user requests [0,USER_TOP]? That sounds daunting and like a potential denial of service.
> > 
> > 
> > The VSpace operations are still only allowed to operate one page at a time, and their implementation looks up the frame cap to validate that something is mapped there.
> 
> @lsf37 My apology here. I just looked at the code; there is check in the kernel to ensure the start and end addresses are in the same page in the VSpace flush function. So, we do not have a denial of service issue. The VSpace invalidate/flush function already calls lookupFrame to do page table walking to ensure the start address is a valid mapping, so it should be easy to include the permission in the resolve_ret structure. The ARMPageClean/Invalidate calls with frame caps do not perform page table walking. But it is straightforward to call lookupFrame to get the permission. Note that if we want to use the AT instructions to check the permission, we need to switch the user vspace root to the target vspace root if necessary.



> 
> > My opinion is that we find a way to place the actual mapping information into the frame cap and as part of decode reject the operation as invalid if the mapping doesn't have write rights. There is space on aarch64 but not on aarch32, we could reduce the ASID space by 1 bit and this should be enough to encode whether the page is mapped with write rights.
> > 
> > * we shouldn't cause an exception for an unexceptional condition that we could avoid if we track the state ourselves.
> > * Longer term we should find a different way of implementing cache maintenance operations so they more accurately reflect the underlying hardware architecture.

**Comment by lsf37** [comment]
So many points :-). I'll make separate comments for each.

>> @kent-mcleod, you have a point that always requiring write for a flush may be too much -- although in the scenario that you described (mapped read-only, but a device can write to it), wouldn't we get the same fault we are getting now? If the hardware requires the page to be mapped with write for this operation, there's not much choice for us apart from requiring it too. Or am I missing something?

> @lsf37 The device is either accessing memory via bypassing the mmu, or through an IOMMU that may have different mapping permissions for the memory.

That's correct, but I didn't mean the device itself. Let's assume the device has separate access to a page, and the user who wants to flush has read access only. Would the user trying to flush in that situation not lead to the same hardware exception we are getting now? If I understand correctly, the user needs write permission in the PDE to run the operation. That means having a separate access right on the seL4 side won't help, unless maybe we remap the PTE to write for the flush operation only and then remap to read-only just after.

**Comment by lsf37** [comment]
> @lsf37 My apology here. I just looked at the code; there is check in the kernel to ensure the start and end addresses are in the same page in the VSpace flush function. So, we do not have a denial of service issue. The VSpace invalidate/flush function already calls lookupFrame to do page table walking to ensure the start address is a valid mapping, so it should be easy to include the permission in the resolve_ret structure. The ARMPageClean/Invalidate calls with frame caps do not perform page table walking. But it is straightforward to call lookupFrame to get the permission.

This sounds promising. If it's restricted to a page already anyway, then it would be perfectly fine to walk the table and check the PTE. I think that's the easiest solution.

**Comment by lsf37** [comment]
Some thoughts on the previous options, mostly in case we want to come back to them for other such situations:

 - taking a kernel exception during the `perform` phase: one thing to take into account is wether the exception is synchronous or not. If it is something async, i.e. we might get the exception slightly later when other operations have been executed already, that wouldn't really be feasible for verification, because we can't predict in the spec how far we might have gotten. Also, the model for the exception/interrupt we do take (in normal current preemption) is that we leave the kernel and restart, whereas here we would want to send a fault message. This is more different to our normal preemption than I had initially realised.

 - preemption: I hadn't though of the fact that this would be preemption during the decode phase. That doesn't really work, because the basis for the previous checks might have changed by the time we complete the check. We'd have to lock down any operation on that VSpace that could potentially interfere. All other cases of preemption we have so far are in the `perform` phase, and the decode provides the conditions for making that lockdown either possible or not necessary. So this is not really a good option, at least it would significantly change how the `decode`/`perform` phases work in the kernel.

- we could probably combine the "take exception" with the "preemption" option in the sense that we do not do preemption in decode (i.e. not do a long check), but do preemption in `perform`, specifically guarding the spots where we could take that kernel exception, make sure nothing asynchronous can happen or at least have a barrier at some point. We'd still have the problem that the execution model is not set up to continue from such an exception. It has to discard the current kernel stack and start something new. Possible, but high verification overhead.

**Comment by yyshen** [comment]
> Some thoughts on the previous options, mostly in case we want to come back to them for other such situations:
> 
> * taking a kernel exception during the `perform` phase: one thing to take into account is wether the exception is synchronous or not. If it is something async, i.e. we might get the exception slightly later when other operations have been executed already, that wouldn't really be feasible for verification, because we can't predict in the spec how far we might have gotten. Also, the model for the exception/interrupt we do take (in normal current preemption) is that we leave the kernel and restart, whereas here we would want to send a fault message. This is more different to our normal preemption than I had initially realised.
> * preemption: I hadn't though of the fact that this would be preemption during the decode phase. That doesn't really work, because the basis for the previous checks might have changed by the time we complete the check. We'd have to lock down any operation on that VSpace that could potentially interfere. All other cases of preemption we have so far are in the `perform` phase, and the decode provides the conditions for making that lockdown either possible or not necessary. So this is not really a good option, at least it would significantly change how the `decode`/`perform` phases work in the kernel.
> * we could probably combine the "take exception" with the "preemption" option in the sense that we do not do preemption in decode (i.e. not do a long check), but do preemption in `perform`, specifically guarding the spots where we could take that kernel exception, make sure nothing asynchronous can happen or at least have a barrier at some point. We'd still have the problem that the execution model is not set up to continue from such an exception. It has to discard the current kernel stack and start something new. Possible, but high verification overhead.

@lsf37 Given that both the vspace-based (start, end) operation (if the start and and are not in the same page, an error is returned) and framecap-based operation only operate on one page for each invocation, the preemption may not be necessary. Another trick is to invalidate translate the user VA to kernel VA and invalidate from the kernel's virtual address, if the caches are physically tagged (likely they are for modern Arm processors). The disadvantage is it could cause some hard to find data corruptions.

**Comment by kent-mcleod** [comment]
> My opinion is that we find a way to place the actual mapping information into the frame cap and as part of decode reject the operation as invalid if the mapping doesn't have write rights. There is space on aarch64 but not on aarch32, we could reduce the ASID space by 1 bit and this should be enough to encode whether the page is mapped with write rights.

I forgot that decode already walks the page tables during decode phase for many operations anyway.  We already check that a frame is mapped into a vspace to validate the ASID, so we could also check the mapping rights during decode and return an error if the mapping doesn't have write attributes.  This could be done for both frame and vspace level invocations right?  Then the kernel would return an error instead of taking a fault.  And we require that the frame is mapped in write to be able to do cache ops for now?

**Comment by lsf37** [comment]
Yes, I think that is a good plan. The only remain question for me is if all architectures require write permission, and for which operations specifically.

**Comment by yyshen** [comment]
> > My opinion is that we find a way to place the actual mapping information into the frame cap and as part of decode reject the operation as invalid if the mapping doesn't have write rights. There is space on aarch64 but not on aarch32, we could reduce the ASID space by 1 bit and this should be enough to encode whether the page is mapped with write rights.
> 
> I forgot that decode already walks the page tables during decode phase for many operations anyway. We already check that a frame is mapped into a vspace to validate the ASID, so we could also check the mapping rights during decode and return an error if the mapping doesn't have write attributes. This could be done for both frame and vspace level invocations right? Then the kernel would return an error instead of taking a fault. And we require that the frame is mapped in write to be able to do cache ops for now?

The page table walking is used by vspace + virtual address operation to find the corresponding physical address and verify the mapping is valid. For operations with a frame cap as input, the page table walking is not performed since related information can be extracted from the frame cap.

**Comment by yyshen** [comment]
> Yes, I think that is a good plan. The only remain question for me is if all architectures require write permission, and for which operations specifically.

A quick check for armv7, copied from ARM 

Cache or branch predictor maintenance operations cannot cause a Permission fault, except that:
• a stage 1 translation table walk performed as part of a cache or branch predictor maintenance operation can
generate a stage 2 Permission fault as described in Stage 2 fault on a stage 1 translation table walk,
Virtualization Extensions on page B3-1402.
• a DCIMVAC issued in Non-secure state that attempts to update date in a location for which it does not have
stage 2 write access can generate a stage 2 Permission fault, as described in Virtualization Extensions
upgrading of maintenance operations on page B2-1286.

For x86, we do not provide similar invalidate/clean calls as we do for Arm.

**Comment by xurtis** [comment]
Seems this was never fixed and we have AOS students hitting it again.

**Comment by lucypa** [comment]
I've run into this issue again and want to restart the conversation so it can be fixed asap

**Comment by kent-mcleod** [comment]
The fix for this would have to be different depending on whether the kernel is managing stage 1 or stage 2 translations.
- stage 1: When the kernel is handling stage 1 translations, it uses the frame's mapped vspace and performs DC IVAC on the supplied vaddr. Fixing this would require looking up the rights of the mapping in the vspace (ie modifying `(lookup_frame_ret = lookupFrame)` to also return the mapping access permissions).  This is likely a fairly straightforward fix if we just error out if the frame isn't mapped writable. (Although what about situations where the frame is mapped readonly, but the frame cap has write rights??)
- stage 2: when the kernel is handling stage 2 translations, it uses a virtual address in the kernel window and relies on the cache being physically indexed. I believe all mappings in the kernel window are write-access and so there shouldn't be a permission fault generated, however if the physical address isn't mapped into the kernel window, then the kernel has no way of performing any cache maintenance operations on the frame either...

**Comment by kent-mcleod** [comment]
So I guess we just need to fix this for the stage 1 variant which would require adding the permission lookup to the `seL4_ARM_Page_Invalidate_Data` invocation, and also checking the permission in the vspace flavor of the invocation.

**Comment by gernotheiser** [comment]
> (Although what about situations where the frame is mapped readonly, but the frame cap has write rights??)
If the user has the rights to write, but choses to use a R/O mapping, then either allowing or preventing the operation seems safe. 
IIUC, the HW would produce a fault if the mapping is R/O. Is the problem that you cannot easily detect this situation proactively when the user has a RW cap?

**Comment by lsf37** [comment]
> > (Although what about situations where the frame is mapped readonly, but the frame cap has write rights??)
> 
> If the user has the rights to write, but choses to use a R/O mapping, then either allowing or preventing the operation seems safe. IIUC, the HW would produce a fault if the mapping is R/O.

Yes, we would want to prevent that fault, because currently the kernel does not expect any faults to happen in kernel mode.

> Is the problem that you cannot easily detect this situation proactively when the user has a RW cap?

It's mostly a question of where to check. Checking the cap is cheap and easy, but not sufficient to prevent the scenario above, so it looks like we need to check the page tables, which is less cheap and easy, but not really a big problem either.

What I'm unclear on is checking entire address ranges (i.e. the VSpace flush operation that has a `[start..end]` va parameter) -- that seems arbitrarily expensive to me, because you'd need to find all page table entries for that address range, which could in theory be an entire fine-grained page table. Maybe that is not actually that much, but it doesn't sound fast.

**Comment by gernotheiser** [comment]
we could do the usual thing of restricting the size of the range supported, to the point of diminishing returns.
In practice, a large range would probably be better done at a full-cache granularity (but that requires special privilege).

**Comment by kent-mcleod** [comment]
The VSpace flush operation already restricts the range to a single frame.  It does this by looking up the frame mapped at the supplied start address and requires that the end address is within the same frame.

**Comment by lsf37** [comment]
Oh, cool, I missed that. Then I think we have a solution that works.

**Comment by lsf37** [comment]
Anybody want to volunteer for putting up a PR?

**Comment by lsf37** [comment]
If nobody gets there before you, yes please. If anyone is starting work on it before end of next week, please comment and coordinate here.

The update will need verification, so I've put that label on, and we'll need to minimise changes for proof impact so we can verify this immediately without waiting for funding. It'd be nice to add a precondition to the corresponding machine interface operation that forces us to prove that we have checked the access rights correctly. We'll have a look at if that is feasible on the verification side.

**Comment by kent-mcleod** [comment]
Something to check when there's an implementation is that there's 4 different mapping types:
VMKernelOnly
VMReadWrite
VMKernelReadOnly
VMReadOnly

So we'd need to decide whether VMKernelOnly or VMKernelReadOnly should be allowed?  I'm guessing that the kernel types should already be guarded by `if (end > USER_TOP) {//       /* Don't let applications flush kernel regions. */` so only VMReadWrite should be allowed.

**Comment by kent-mcleod** [comment]
We need to add some tests to sel4test for this.  In particular we probably need to check for each arch that the following table is what's implemented:

| Mapping type\ cache op| Clean_Data | Invalidate_Data | CleanInvalidate_Data | Unify_Instruction |
| - | - | - | - | - |
| Invalid Mapping | :x:  | :x:  | :x: | :x:  |
| VMKernelOnly | :x:  | :x:  | :x: | :x:  |
| VMReadWrite | :heavy_check_mark: | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |
| VMKernelReadOnly | :x:  | :x:  | :x: | :x:  |
| VMReadOnly | :heavy_check_mark: | :x:  | :heavy_check_mark: | :heavy_check_mark: |

I think at the moment, a frame cap with no rights (assuming that CnodeMint allows you to create a frame cap with neither write or read rights) can be mapped (which will create a VMKernelOnly mapping) inaccessible from normal user space accesses, but the kernel will still execute all of the cache ops as currently just possessing the frame cap is considered sufficient authority.  That is to say, I think the current implementation for aarch32 and aarch64 would all :heavy_check_mark: for the cache ops by frame cap, and :x: for the first row but otherwise all :heavy_check_mark: for cache ops by vspace.

**Comment by yyshen** [comment]
> I can pick it up, but won't be able to work on it till end of next week at the earliest.

would you do the page table walk or use the AT instruction?

**Comment by lsf37** [comment]
> > I can pick it up, but won't be able to work on it till end of next week at the earliest.
> 
> would you do the page table walk or use the AT instruction?

Given that we already have code for it I think it would make sense to walk in software. Let me double-check where that would be.

I'm not entirely clear on whether we have this problem in both, aarch32 and aarch64, or just in aarch64. @yyshen's comment for armv7

 > a DCIMVAC issued in Non-secure state that attempts to update date in a location for which it does not have
stage 2 write access can generate a stage 2 Permission fault, as described in Virtualization Extensions
upgrading of maintenance operations on page B2-1286.

seems to say that this is an issue for both. Do I understand this correctly?

**Comment by kent-mcleod** [comment]
We may want to split part of the discussion off into an additional ticket because there's the specific issue and then potentially a more general issue that we probably should address but has lower significance.
 
There's the specific problem: "seL4_ARM_Page_Invalidate_Data doesn't check if the page is writeable" where the consequence is a kernel write fault:
- Fixing this likely only requires changes to non-verified configurations (EL1 AArch64).
- Verified configurations (aarch64-hyp) don't trigger a fault as they don't flush using the callers vspace and instead have to flush using the kernel's vspace which has to have a writeable mapping as it's the kernel window mapping.
- aarch32 doesn't require the address to be write-able and so the kernel doesn't need the check.
- I think the part about aarch32 stage-2 permission faults on DCIMVAC instructions shouldn't affect the kernel because it's performing the op through the hyp-mode address space that doesn't have stage 2 translation.

Then there's the more general problem: "None of the cache invocations check the existing page mapping rights or page cap rights" were the consequence is that a frame cap or vspace cap gives you ambient authority to perform cache ops on the addressable memory regardless of the cap rights.

**Comment by gernotheiser** [comment]
> Then there's the more general problem: "None of the cache invocations check the existing page mapping rights or page cap rights" were the consequence is that a frame cap or vspace cap gives you ambient authority to perform cache ops on the addressable memory regardless of the cap rights.

Not sure what the problem is here, as you aren't doing an operation on someone else's behalf. If you have a W cap, then no matter whether your present mapping is writeable or not, you have the power to make it writable, and I don't see why you shouldn't be allowed to invalidate data you can modify if you want to.
Maybe I misunderstand what you're saying.

**Comment by gernotheiser** [comment]
> aarch32 doesn't require the address to be write-able and so the kernel doesn't need the check.

That seems a problem to me. If a page is shared between A and B, where A has a R/O mapping and B a RW mapping, then A should not be allowed to modify any of its data. However, if B modifies it, and while it's still dirty in the cache, A invalidates it, then A undoes B's writes without having permission to do so.

**Comment by Ivan-Velickovic** [comment]
> We need to add some tests to sel4test for this. In particular we probably need to check for each arch that the following table is what's implemented:
> Mapping type\ cache op 	Clean_Data 	Invalidate_Data 	CleanInvalidate_Data 	Unify_Instruction
> Invalid Mapping 	❌ 	❌ 	❌ 	❌
> VMKernelOnly 	❌ 	❌ 	❌ 	❌
> VMReadWrite 	✔️ 	✔️ 	✔️ 	✔️
> VMKernelReadOnly 	❌ 	❌ 	❌ 	❌
> VMReadOnly 	✔️ 	❌ 	✔️ 	✔️
> 
> I think at the moment, a frame cap with no rights (assuming that CnodeMint allows you to create a frame cap with neither write or read rights) can be mapped (which will create a VMKernelOnly mapping) inaccessible from normal user space accesses, but the kernel will still execute all of the cache ops as currently just possessing the frame cap is considered sufficient authority. That is to say, I think the current implementation for aarch32 and aarch64 would all ✔️ for the cache ops by frame cap, and ❌ for the first row but otherwise all ✔️ for cache ops by vspace.

I can work on this and make a PR.

**Comment by kent-mcleod** [comment]
@gernotheiser, in the case where there's a frame cap with no access rights, a call to `ARMPageMap` will result in a mapping with KernelOnly rights[1] . If you then attempt any of the above mentioned cache maintenance operations, the kernel will just do it and it'll succeed because of the existence of the kernel mapping. This should be considered a bug.

[1] (in the architecture page table entries, the access rights field (AP[2:1]) has one bit for selecting user mode or privileged-mode access, and second bit for selecting Read-write or read-only access)

**Comment by kent-mcleod** [comment]
https://github.com/seL4/seL4/pull/1041 would be the smallest change to introduce the check I think. (It doesn't use the AT instructions)

**Comment by kent-mcleod** [comment]
> I can work on this and make a PR.

That'd be amazing

**Comment by Ivan-Velickovic** [comment]
At the end of this we should document the relevant permissions required for the cache operations. Given that the operations can also be done from user-space, a section on cache management in the manual might be warranted.
